samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -200,80 +275,94 @@ SamzaContainer createSamzaContainer(String processorId, JobModel jobModel) {
-              }
-              if (!shutdownComplete) {
-                LOGGER.warn("Container: {} shutdown was unsuccessful. Stopping the stream processor: {}.", container, processorId);
-                container = null;
-                stop();
-              } else {
-                LOGGER.info("Container: {} shutdown completed for stream processor: {}.", container, processorId);
-              }
+          if (state == State.STARTED || state == State.RUNNING) {
+            state = State.IN_REBALANCE;
+            LOGGER.info("Job model expired. Shutting down the container: {} of stream processor: {}.", container, processorId);
+            boolean hasContainerShutdown = stopSamzaContainer();
+            if (!hasContainerShutdown) {
+              LOGGER.warn("Container: {} shutdown was unsuccessful. Stopping the stream processor: {}.", container, processorId);
+              state = State.STOPPING;
+              jobCoordinator.stop();
             } else {
-              LOGGER.info("Container: {} of the stream processor: {} is not running.", container, processorId);
+              LOGGER.info("Container: {} shutdown completed for stream processor: {}.", container, processorId);
             }
           } else {
-            LOGGER.info("Container is not instantiated for stream processor: {}.", processorId);
+            LOGGER.info("Container: {} of the stream processor: {} is not running.", container, processorId);
           }
         }
       }
 
       @Override
       public void onNewJobModel(String processorId, JobModel jobModel) {
         synchronized (lock) {
-          jcContainerShutdownLatch = new CountDownLatch(1);
-          container = createSamzaContainer(processorId, jobModel);
-          container.setContainerListener(new ContainerListener());
-          LOGGER.info("Starting the container: {} for the stream processor: {}.", container, processorId);
-          executorService.submit(container::run);
+          if (state == State.IN_REBALANCE) {
+            containerShutdownLatch = new CountDownLatch(1);
+            container = createSamzaContainer(processorId, jobModel);
+            container.setContainerListener(new ContainerListener());
+            LOGGER.info("Starting the container: {} for the stream processor: {}.", container, processorId);
+            executorService.submit(container);
+          } else {
+            LOGGER.info("Ignoring onNewJobModel invocation since the current state is {} and not {}.", state, State.IN_REBALANCE);
+          }
         }
       }
 
       @Override
       public void onCoordinatorStop() {
-        if (executorService != null) {
+        synchronized (lock) {
           LOGGER.info("Shutting down the executor service of the stream processor: {}.", processorId);
+          stopSamzaContainer();
           executorService.shutdownNow();
+          state = State.STOPPED;
         }
-        if (processorListener != null) {
```

Comment:
if you don't want to check processListener for null here, then add this check in constructor (or assert)

Reply:
Makes sense. Made the change.